### PR TITLE
Added cache for requests with sync/async storage

### PR DESCRIPTION
**Pull request by minorsetback**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,3 +1,24 @@
+interface AsyncStorage {
+    getItem: (key: string) => Promise<string | undefined | null>;
+    setItem: (key: string, value: string) => Promise<unknown>;
+    removeItem: (key: string) => Promise<void>;
+}
+
+interface AsyncSourceConfig {
+    storage?: CacheStorage;
+    time?: number;
+    prefix?: string;
```
Comment (SergiiKudriashov) [inline src/index.ts:10]: Prefix should be named semantically to provide understating of what this field is stand for 
Suggestion : `cacheKeyPrefix`

File: src/index.ts [hunk 2]
```diff
@@ -1,3 +1,24 @@
...
+}
```
Comment (SergiiKudriashov) [inline src/index.ts:11]: add `isUpdateCache` flag that should provide ability to send/ignore request

File: src/index.ts [hunk 3]
```diff
@@ -11,15 +32,71 @@ class AsyncSource<T> {
...
+        } else {
+            const {
+                debounceTime = 100,
+                cacheKey,
+                cacheTime = AsyncSource.defaultCacheTime,
+                storage = AsyncSource.defaultStorage
+            } = debounceTimeOrConfig;
+
+            this.debounceTime = debounceTime;
+            this.cacheKey = cacheKey ?  `${AsyncSource.prefix}-${cacheKey}` : undefined;
```
Comment (SergiiKudriashov) [inline src/index.ts:83]: cacheKey and this.cacheKey should be named differently

File: src/index.ts [hunk 4]
```diff
@@ -73,10 +150,90 @@ class AsyncSource<T> {
...
+                const isExpired = Date.now() - timestamp > this.cacheTime;
+                return isExpired ? null : (data as PromiseResult<T>);
+            } catch (error) {
+                this.onError(error as Error);
+            }
+        }
+        return null;
+    }
+
+    private setArgs(args: Array<any>) {
```
Comment (SergiiKudriashov) [inline src/index.ts:198]: should be renamed to provide more semantic 
Suggestion: 
```
private setLastrequestParams(requestParams: Array<any>) {
```

File: src/index.ts [hunk 5]
```diff
@@ -73,10 +150,90 @@ class AsyncSource<T> {
...
+        if (args) {
+            try {
+                this.args = JSON.stringify(args);
+            } catch (error) {
+                this.onError(error as Error);
```
Comment (SergiiKudriashov) [inline src/index.ts:203]: `this.onError(error as Error);`

Should be changed to some kind of console warning
Suggestions: `Cache saving error ${url} ${params}`

File: src/index.ts [hunk 6]
```diff
@@ -73,10 +150,90 @@ class AsyncSource<T> {
...
+
+        const cacheValue = {
+            data,
+            timestamp: Date.now()
+        };
+
+        try {
+            new Promise((resolve) => resolve(this.storage?.setItem(cacheKey, JSON.stringify(cacheValue))));
+        } catch (error) {
+            this.onError(error as Error);
```
Comment (SergiiKudriashov) [inline src/index.ts:221]: this.onError(error as Error);

Should be changed to some kind of console warning
Suggestions: Cache saving error ${url} ${params}

File: src/index.ts [hunk 7]
```diff
@@ -73,10 +150,90 @@ class AsyncSource<T> {
...
+        }
+    }
+
+    private async request(args: Array<any>, successHandler?: (response: T) => void, isImmediate?: boolean): Promise<void> {
+        this.setArgs(args);
```
Comment (SergiiKudriashov) [inline src/index.ts:226]: 1. Add additional check for enabled state to skip current cache logic when cache is not configured
2. Move current after creating requestId

File: src/index.ts
Comment (SergiiKudriashov) [inline src/index.ts:1]: Add unit tests

Update readme.md

File: src/index.ts [hunk 8]
```diff
@@ -73,22 +135,110 @@ class AsyncSource<T> {
...
+                requestParamsHash = JSON.stringify(requestParams);
+            } catch (error) {
+                console.warn({ message: `Request params saving error cacheKey:${baseCacheKey}`, params: requestParams, error });
+            }
+        }
+
+        this.cacheKey = requestParamsHash ? `${baseCacheKey}-${requestParamsHash}` : baseCacheKey;
     }
 
-    // Core request method
```
Comment (prochorz) [inline src/index.ts:78]: save this comment

File: src/index.ts [hunk 9]
```diff
@@ -73,22 +135,110 @@ class AsyncSource<T> {
...
+        }
+    }
+
+    private async request(args: Array<any>, successHandler?: (response: T) => void, isImmediate?: boolean): Promise<void> {
         this.isRequestPending = true;
 
         const requestId = await this.createRequestId(isImmediate);
         if (!this.isLastRequest(requestId)) return;
 
+        if (this.isCacheEnabled) {
```
Comment (prochorz) [inline src/index.ts:220]:
```
if (this.isCacheEnabled && this.isLastRequest(requestId)) {
```
It seems it should be like this.

Comment (minorsetback) [reply src/index.ts:220]: `if (!this.isLastRequest(requestId)) return;` but we have this? it's the same check i think

File: src/index.ts [hunk 10]
```diff
@@ -73,22 +135,113 @@ class AsyncSource<T> {
...
+            }
+        }
+
+        this.cacheKey = requestParamsHash ? `${baseCacheKey}-${requestParamsHash}` : baseCacheKey;
+    }
+
+    private removeCachedData() {
+        try {
+            if (this.cacheKey) {
+                this.cacheStorage?.removeItem?.(this.cacheKey);
```
Comment (prochorz) [inline src/index.ts:173]: await

File: src/index.ts [hunk 11]
```diff
@@ -11,15 +33,55 @@ class AsyncSource<T> {
...
+                cacheStorage = AsyncSource.defaultStorage,
+                isUpdateCache = true
+            } = debounceTimeOrConfig;
+
+            this.debounceTime = debounceTime;
+            this.isCacheEnabled = Boolean(requestCacheKey);
+            this.requestCacheKey = requestCacheKey;
+            this.cacheTime = cacheTime;
+            this.cacheStorage = cacheStorage;
+            this.isUpdateRequestCache = isUpdateCache;
```
Comment (prochorz) [inline src/index.ts:70]: `this.isUpdateCache = isUpdateCache;`

File: src/index.ts [hunk 12]
```diff
@@ -73,22 +135,113 @@ class AsyncSource<T> {
...
+            } catch (error) {
+                console.warn({ message: `Request params saving error cacheKey:${baseCacheKey}`, params: requestParams, error });
+            }
+        }
+
+        this.cacheKey = requestParamsHash ? `${baseCacheKey}-${requestParamsHash}` : baseCacheKey;
+    }
+
+    private async removeCachedData() {
+        try {
```
Comment (prochorz) [inline src/index.ts:171]: `if (!this.cacheKey) return;`
To make it the same

**Pull request merged**